### PR TITLE
Llama Index serde and pyfunc

### PR DESCRIPTION
**Pull request by michael-berk**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/michael-berk/mlflow/pull/11674?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11674/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11674
```

</p>
</details>

### What changes are proposed in this pull request?

Commits `10` (ready for review)
* Add serialize objects util
* Add tests for serialize objects util
* Add save/load functionality
* Add simple e2e test on save/load
* Add subs for signature inference - not ready for review

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [x] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Official flavor support for llama index. 

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: mlflow/llama_index/__init__.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+    import llama_index
+
+    _validate_env_arguments(conda_env, pip_requirements, extra_pip_requirements)
+
+    path = os.path.abspath(path)
+    _validate_and_prepare_target_save_path(path)
+    code_dir_subpath = _validate_and_copy_code_paths(code_paths, path)
+
+    if signature is None and input_example is not None:
+        wrapped_model = _LlamaIndexModelWrapper(index)
```
Comment (BenWilson2) [inline mlflow/llama_index/__init__.py:103]: Does the input / output schema change for the type of engine that is specified?

File: mlflow/llama_index/__init__.py [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+    if signature is not None:
+        signature = validate_and_resolve_signature(index, signature)
+        mlflow_model.signature = signature
+    if input_example is not None:
+        _save_example(mlflow_model, input_example, path)
+    if metadata is not None:
+        mlflow_model.metadata = metadata
+
+    engine_kwargs_path = os.path.join(path, _ENGINE_KWARGS_DIRECTORY)
+    serialize_engine_kwargs_to_json(engine_kwargs, engine_kwargs_path)
```
Comment (BenWilson2) [inline mlflow/llama_index/__init__.py:167]: Use model_config for this :)

File: mlflow/llama_index/__init__.py [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+
+    from llama_index.core import Settings
+
+    settings_path = os.path.join(path, _SETTINGS_DIRECTORY)
+    serialize_settings_to_json(Settings, settings_path)
```
Comment (BenWilson2) [inline mlflow/llama_index/__init__.py:172]: This needs a dev note comment explaining what is happening here. Why we're accessing a singleton and what we're extracting from it.

File: mlflow/llama_index/__init__.py [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+
+    model_data_path = path
```
Comment (BenWilson2) [inline mlflow/llama_index/__init__.py:174]: useless assign?

File: mlflow/llama_index/__init__.py [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+    _save_model(index, model_data_path)
+
+    flavor_conf = {"engine_type": engine_type}
```
Comment (BenWilson2) [inline mlflow/llama_index/__init__.py:177]: Do we need validation of engine_type?

File: mlflow/llama_index/__init__.py [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+    if _should_add_pyfunc_to_model(index):
+        if mlflow_model.signature is None:
```
Comment (BenWilson2) [inline mlflow/llama_index/__init__.py:179]: Wasn't this done above?

File: mlflow/llama_index/__init__.py [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+    pyfunc.add_to_model(
+        mlflow_model,
+        loader_module="mlflow.llama_index",
+        conda_env=_CONDA_ENV_FILE_NAME,
+        python_env=_PYTHON_ENV_FILE_NAME,
+        code=code_dir_subpath,
+    )
+    mlflow_model.add_flavor(
+        FLAVOR_NAME,
+        llama_index_version=llama_index.core.__version__,
```
Comment (BenWilson2) [inline mlflow/llama_index/__init__.py:195]: Let's add a utility for this (getting the version)

File: mlflow/llama_index/__init__.py [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+        code=code_dir_subpath,
+        **flavor_conf,
+    )
+    if size := get_total_file_size(path):
```
Comment (BenWilson2) [inline mlflow/llama_index/__init__.py:199]: Validate with the new Resources object type to determine if this is the correct way of doing this now. Parameters that are sent to model serving for validating instance type for serving considerations should conform to the standard interface.

Comment (BenWilson2) [inline mlflow/llama_index/__init__.py:199]: See how langchain does Resource definition and copy that for submitting deployment metadata for model serving to use (we will need to make changes in serving for this flavor to consume this metadata obj)

File: mlflow/llama_index/__init__.py [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+
+    _PythonEnv.current().to_yaml(os.path.join(path, _PYTHON_ENV_FILE_NAME))
+
+
+@format_docstring(LOG_MODEL_PARAM_DOCS.format(package_name=FLAVOR_NAME))
+def log_model(
+    index,
+    artifact_path: str,
+    engine_type: str,
+    engine_kwargs: Optional[Dict[str, any]] = None,
```
Comment (BenWilson2) [inline mlflow/llama_index/__init__.py:231]: this should be `model_config`

File: mlflow/llama_index/__init__.py [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+            path will be created.
+
+    Returns:
+        A ``llama_index`` index object.
+    """
+    local_model_path = _download_artifact_from_uri(artifact_uri=model_uri, output_path=dst_path)
+    flavor_conf = _get_flavor_configuration(model_path=local_model_path, flavor_name=FLAVOR_NAME)
+
+    settings_path = os.path.join(local_model_path, _SETTINGS_DIRECTORY)
+    # Settings is a singleton, so it does not need to be passed
```
Comment (BenWilson2) [inline mlflow/llama_index/__init__.py:341]:
```suggestion
    # Settings is a singleton
```

File: mlflow/llama_index/__init__.py [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+    _ = deserialize_json_to_settings(settings_path)
+
+    return _load_model(local_model_path, flavor_conf)
+
+
+def _load_pyfunc(path):
+    engine_kwargs_path = os.path.join(path, _ENGINE_KWARGS_DIRECTORY)
+    engine_kwargs = deserialize_json_to_engine_kwargs(engine_kwargs_path)
```
Comment (BenWilson2) [inline mlflow/llama_index/__init__.py:349]: read from model_config if defined

File: mlflow/llama_index/__init__.py [hunk 12]
```diff
@@ -0,0 +1,385 @@
...
+
+    flavor_conf = _get_flavor_configuration(model_path=path, flavor_name=FLAVOR_NAME)
+    engine_type = flavor_conf["engine_type"]
+    return _LlamaIndexModelWrapper(_load_model(path, flavor_conf), engine_type, engine_kwargs)
```
Comment (BenWilson2) [inline mlflow/llama_index/__init__.py:353]: extract engine_type and engine_kwargs from model_config by popping. What other args are available for instantiating that we might use model_config kwargs entries for? How will these be handled?

File: mlflow/llama_index/serialize_objects.py [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+from typing import Callable, Dict, Set, Tuple
+
+from llama_index.core import PromptTemplate, Settings
+
+_logger = logging.getLogger(__name__)
+
+_SUPPORTED_SERIALIZATION_METHODS = ["to_dict", "dict"]
+
+
+def _extract_constructor_from_object(o: object) -> Callable:
```
Comment (BenWilson2) [inline mlflow/llama_index/serialize_objects.py:14]: if callable(obj) will replace most of this logic :)

File: mlflow/llama_index/serialize_objects.py [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+            "be concerted to a callable."
+        )
+    else:
+        raise AttributeError(
+            f"Object {o} does not have __class__ attribute and thereby cannot be "
+            "converted to callable."
+        )
+
+
+def _get_dict_method_if_exists(o: object) -> Dict[str, any]:
```
Comment (BenWilson2) [inline mlflow/llama_index/serialize_objects.py:30]: Let's change this to isinstance(BaseClass) for the types that we are certain can be serialized :)

File: mlflow/llama_index/serialize_objects.py [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+    for method in _SUPPORTED_SERIALIZATION_METHODS:
+        if hasattr(o, method):
+            return getattr(o, method)()
+
+    raise AttributeError(f"Object {o} does not have a supported deserialization method.")
+
+
+def _get_kwargs(c: Callable) -> Tuple[Set]:
```
Comment (BenWilson2) [inline mlflow/llama_index/serialize_objects.py:38]: Let's verify what LlamaIndex does here with unexpected args. If it throws, we don't need this. If it creates an impossible to diagnose error, we'll need some validation. But for now, we don't need this validation.

File: mlflow/llama_index/serialize_objects.py [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+
+        if not hasattr(module, class_name):
+            raise ValueError(f"Module {module} does not have {class_name}")
+
+    return f"{module_name}.{class_name}"
+
+
+def _sanitize_api_key(d: Dict[str, str]) -> Dict[str, str]:
+    # TODO: drop API key and make a note that it must be in an enviornment variable
+    return d
```
Comment (BenWilson2) [inline mlflow/llama_index/serialize_objects.py:74]: after updating this, make sure not to use shorthand variable names like this.

File: mlflow/llama_index/serialize_objects.py [hunk 17]
```diff
@@ -0,0 +1,224 @@
...
+def _sanitize_kwargs(o: object, o_state_as_dict: Dict[str, any]) -> Dict[str, any]:
+    """
+    Sanitize the object kwargs by...
+    1. Asserting that all required kwargs exist in the object state dict.
+    2. Dropping all state that is not an argument in the object constructor signature.
+    """
+    o_callable = _extract_constructor_from_object(o)
+    required_kwargs, optional_kwargs = _get_kwargs(o_callable)
+    missing_kwargs = set(required_kwargs) - set(o_state_as_dict.keys())
+    if len(missing_kwargs) > 0:
```
Comment (BenWilson2) [inline mlflow/llama_index/serialize_objects.py:104]: Python does this for you.

File: mlflow/llama_index/serialize_objects.py [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+
+    if isinstance(module, PromptTemplate):
+        return _construct_prompt_template_object(module, kwargs)
+    else:
+        # Generic object that can be constructed from class __init__ method
+        callable = getattr(module, class_name)
+        return callable(**kwargs)
+
+
+def _deserialize_json_to_dict_of_objects(path: str) -> Dict[str, any]:
```
Comment (BenWilson2) [inline mlflow/llama_index/serialize_objects.py:172]: rename this to something that is intuitive.

File: mlflow/llama_index/serialize_objects.py [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+            )
+        else:
+            to_serialize.update({k: object_json})
+
+    with open(path, "w") as f:
+        json.dump(to_serialize, f, indent=2)
+
+
+def serialize_settings_to_json(settings: Settings, path: str) -> None:
+    _serialize_dict_of_objects_to_json(settings.__dict__, path)
```
Comment (BenWilson2) [inline mlflow/llama_index/serialize_objects.py:207]: Why would we not just call this directly where this is being used?

Comment (BenWilson2) [inline mlflow/llama_index/serialize_objects.py:207]:
```suggestion
def serialize_settings_to_json(path: str) -> None:
    from llama_index.core import Settings

    _serialized_dict_of_objects_to_json(Settings.__dict__, path)
```

File: mlflow/llama_index/serialize_objects.py [hunk 20]
```diff
@@ -0,0 +1,224 @@
...
+
+
+def deserialize_json_to_settings(path: str) -> Settings:
+    settings_dict = _deserialize_json_to_dict_of_objects(path)
+
+    for k, v in settings_dict.items():
+        setattr(Settings, k, v)
+
+    return Settings
```
Comment (BenWilson2) [inline mlflow/llama_index/serialize_objects.py:216]: singleton. It's global.

File: mlflow/llama_index/serialize_objects.py [hunk 21]
```diff
@@ -0,0 +1,224 @@
...
+
+    with open(path, "w") as f:
+        json.dump(to_serialize, f, indent=2)
+
+
+def serialize_settings_to_json(settings: Settings, path: str) -> None:
+    _serialize_dict_of_objects_to_json(settings.__dict__, path)
+
+
+def deserialize_json_to_settings(path: str) -> Settings:
```
Comment (BenWilson2) [inline mlflow/llama_index/serialize_objects.py:210]: Remove return type

File: mlflow/llama_index/serialize_objects.py [hunk 22]
```diff
@@ -0,0 +1,224 @@
...
+    settings_dict = _deserialize_json_to_dict_of_objects(path)
+
+    for k, v in settings_dict.items():
+        setattr(Settings, k, v)
+
+    return Settings
+
+
+def serialize_engine_kwargs_to_json(engine_kwargs: Dict[str, object], path: str) -> None:
```
Comment (BenWilson2) [inline mlflow/llama_index/serialize_objects.py:219]: This handled by `model_config` serialization.

File: mlflow/llama_index/serialize_objects.py [hunk 23]
```diff
@@ -0,0 +1,224 @@
...
+    _serialize_dict_of_objects_to_json(engine_kwargs, path)
+
+
+def deserialize_json_to_engine_kwargs(path: str) -> Dict[str, object]:
```
Comment (BenWilson2) [inline mlflow/llama_index/serialize_objects.py:223]: same as above.

File: mlflow/llama_index/signature.py [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+
+def generate_signature_output(index, data, model_config=None, flavor_config=None, params=None):
+    """
+    PLACEHOLDER
+    """
+    # Lazy import to avoid circular dependencies. Ideally we should move _LlamaIndexModelWrapper
+    # out from __init__.py to avoid this.
+    from mlflow.llama_index import _LlamaIndexModelWrapper
+
+    return _LlamaIndexModelWrapper(
```
Comment (BenWilson2) [inline mlflow/llama_index/signature.py:24]: However, a utility function within init that constructs this can avoid this import. Build and instance of _LlamaIndexModelWrapper via a private func within the init.py

File: mlflow/llama_index/signature.py [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+        index=index, model_config=model_config, flavor_config=flavor_config
+    ).predict(data, params=params)
+
+
+def infer_signature_from_input_example(
+    index, example=None, model_config=None, flavor_config=None
+) -> ModelSignature:
+    """
```
Comment (BenWilson2) [inline mlflow/llama_index/signature.py:32]: We'll need to support the standard OpenAI specific input / output structures ChatCompletions, etc... for a unified llm interface

File: mlflow/llama_index/__init__.py [hunk 26]
```diff
@@ -348,32 +354,158 @@ def _load_pyfunc(path, model_config: Optional[Dict[str, Any]] = None):
...
+            prediction_list = [
+                self._format_predict_input_chat(d) for d in data.to_dict(orient="records")
+            ]
+        else:
+            raise ValueError(
+                f"Unsupported input type: {type(data)}. It must be one of "
+                "[str, dict, list, numpy.ndarray, pandas.DataFrame]"
+            )
+
+        if len(prediction_list) == 1:
```
Comment (B-Step62) [inline mlflow/llama_index/__init__.py:452]: nit: this holds input not prediction result so might be a bit mislearing

File: mlflow/llama_index/__init__.py [hunk 27]
```diff
@@ -348,32 +354,158 @@ def _load_pyfunc(path, model_config: Optional[Dict[str, Any]] = None):
...
+                payload = next(value for value in data.values())
+                return self._safe_convert_to_chat_messages(payload)
+
+            elif len(data) > 1:
+                # Assume that any kwarg value that's a sequence of strings with a colon
+                # should be converted to a list of ChatMessage objects.
+                return {k: self._safe_convert_to_chat_messages(v) for k, v in data.items()}
+        elif isinstance(data, (list, np.ndarray)):
+            prediction_list = [self._format_predict_input_chat(d) for d in data]
+        elif isinstance(data, pd.DataFrame):
```
Comment (B-Step62) [inline mlflow/llama_index/__init__.py:442]: Can we check what are the possible output type from `_convert_llm_input_data`? From a quick read, seems `np.array` and `pd.DataFrame` are not returned.

File: mlflow/llama_index/__init__.py [hunk 28]
```diff
@@ -348,32 +354,158 @@ def _load_pyfunc(path, model_config: Optional[Dict[str, Any]] = None):
...
+                "{_SUPPORTED_ENGINES}"
+            )
+
+    def __init__(self, index, engine_type: str, model_config: Optional[Dict[str, Any]] = None):
+        self.index = index
+        self.model_config = model_config or {}
+        self.engine_type = engine_type
+        self.predict_callable = self._build_engine_method()
+
+    def _format_predict_input_query_and_retriever(self, data):
```
Comment (B-Step62) [inline mlflow/llama_index/__init__.py:376]: Is it possible to make this function only accept single data and outputs `QueryBundle` only? Since the prediction is made one by one, I think it's fine to do formatting one by one too, and I believe it makes this function simpler?

This is introspection from LangChain experience, having multiple possible data types (e.g. batch vs single) flying within the data flow makes it super hard to reason about the logic and test with high coverage.

File: mlflow/llama_index/__init__.py [hunk 29]
```diff
@@ -348,32 +354,158 @@ def _load_pyfunc(path, model_config: Optional[Dict[str, Any]] = None):
...
+        elif isinstance(data, dict):
+            if len(data) == 1:
+                # Assume that this value should be treated as a positional argument
+                payload = next(value for value in data.values())
+                return self._safe_convert_to_chat_messages(payload)
+
+            elif len(data) > 1:
+                # Assume that any kwarg value that's a sequence of strings with a colon
+                # should be converted to a list of ChatMessage objects.
+                return {k: self._safe_convert_to_chat_messages(v) for k, v in data.items()}
```
Comment (B-Step62) [inline mlflow/llama_index/__init__.py:439]: What is the corresponding LlamaIndex invocation signature for this case? What will be the key for each chat message?

File: mlflow/llama_index/__init__.py [hunk 30]
```diff
@@ -348,32 +354,158 @@ def _load_pyfunc(path, model_config: Optional[Dict[str, Any]] = None):
...
+        # "`role`: `content`", for example "system: You are a helpful bot".
+        if isinstance(data, (list, tuple)) and all(":" in message for message in data):
+            return [ChatMessage.from_str(*(message.split(": ", 1)[::-1])) for message in data]
+
+        return data
+
+    def _format_predict_input_chat(self, data):
+        """
+        Chat engines can have a variety of signatures. The two primary standards are:
+        1. chat(str, Sequence[ChatMessage])
```
Comment (B-Step62) [inline mlflow/llama_index/__init__.py:420]: Q: in this case, the former is the question and the latter is the chat history?

File: mlflow/llama_index/__init__.py [hunk 31]
```diff
@@ -348,32 +354,158 @@ def _load_pyfunc(path, model_config: Optional[Dict[str, Any]] = None):
 
 
 class _LlamaIndexModelWrapper:
```
Comment (B-Step62) [inline mlflow/llama_index/__init__.py:356]: Random idea: I found almost all lines in class have conditioned by engine type. What about having subclasses like `_LlamaIndexChatEngineWrapper`?

File: mlflow/llama_index/__init__.py (hunk 30 above)
Comment (michael-berk) [reply mlflow/llama_index/__init__.py:420]: Correct

File: mlflow/llama_index/__init__.py (hunk 31 above)
Comment (michael-berk) [reply mlflow/llama_index/__init__.py:356]: Genius

File: mlflow/llama_index/__init__.py (hunk 28 above)
Comment (michael-berk) [reply mlflow/llama_index/__init__.py:376]: Really great point!

File: mlflow/llama_index/__init__.py (hunk 29 above)
Comment (michael-berk) [reply mlflow/llama_index/__init__.py:439]: There are several signatures, both in terms of number of parameters and the names of the parameters. There is also often a `**kwargs` in the signature. Some examples:
* `def chat(self, messages: Sequence[ChatMessage], **kwargs: Any) -> ChatResponse:`
* `def chat(self, message: str, chat_history: Optional[List[ChatMessage]] = None) -> AgentChatResponse:`
* `def chat(self, query: str, chat_history: Optional[List[ChatMessage]] = None):`

[Search term in llama_index](https://github.com/search?q=repo%3Arun-llama%2Fllama_index+%22def+chat%22+-path%3Alegacy&type=code)

Here's the thought process
* All signatures accept 1 positional argument, so if the dict has `len == 1`, it should be treated as positional
* All signatures require a message history formatted as `List[ChatMessage]`. The user must pass a list of string `[str(ChatMessage(role='user', content='xyz'))]` (llama_index implements a `__repr__` in `ChatMessage`). Then, we parse this via `_safe_convert_to_chat_messages`.

Would appreciate any suggestions on how to make this more robust - it's a start but is not very dynamic 🙂

Comment (B-Step62) [reply mlflow/llama_index/__init__.py]: Thanks for the explanation, @michael-berk !!

> All signatures accept 1 positional argument, so if the dict has len == 1, it should be treated as positional

Hmm, the `message: str` and `query: str` is not the stringified chat message, but a single question? So the input will look like:

```
model.predict({
    "message": "What is MLflow?"
    "chat_history": [
        {
            "role": "system",
            "content": "xxx",
        },
        {
            "role": "user",
            "content": "yyy",
        }
    ]
```

Also here `chat_history` is passed as list of dictionary, not a list of `ChatMessage`. We need to support both, because in model serving, the request payload is always JSON serialized.

> The user must pass a list of string [str(ChatMessage(role='user', content='xyz'))]

I don't think we can ask that favor to users.... the same input example has to work for  models loaded by `mlflow.llamaindex.load_model()` and `mlflow.pyfunc.load_model()`. Also `str` conversion is not reliable in general, like they (llamaindex) can change `__repr__` implementation to whatever convenient, e.g. omits some field.

My suggestion here is to process the `query`/`message` and other messages separately. What do you think??🤔

```
   messages = []
   if user_query := data.get("query") or data.get("message"):
        messages.append(ChatMessage(role="user", content=user_query)
   if history := data.get("chat_history"):
       messages.extend([ChatMessage(**msg) for msg in history])
```

Comment (michael-berk) [reply mlflow/llama_index/__init__.py]: Nice! Resolved in [this commit](https://github.com/mlflow/mlflow/pull/11674/commits/99478107bb4a3c488794d8139223c20718112d17)

I've added all kwargs that currently exist in llama_index for both `query` and `message_history`.

File: mlflow/llama_index/__init__.py [hunk 32]
```diff
@@ -0,0 +1,338 @@
...
+    # the object that are not serializable, we log a warning.
+    settings_path = os.path.join(path, _SETTINGS_DIRECTORY)
+    serialize_settings(settings_path)
+
+    _save_model(index, path)
+
+    if not model_config:
+        model_config = {"engine_type": engine_type}
+    elif isinstance(model_config, dict):
+        model_config.update({"engine_type": engine_type})
```
Comment (B-Step62) [inline mlflow/llama_index/__init__.py:172]: Shall we raise an exception when `model_config` is not dictionary?

File: mlflow/llama_index/pyfunc_wrapper.py [hunk 33]
```diff
@@ -0,0 +1,159 @@
...
+
+    def _do_inference(self, input, params: Optional[Dict[str, Any]]) -> Dict:
+        """
+        Perform engine inference on a single engine input e.g. not an iterable of
+        engine inputs. The engine inputs must already be preprocessed/cleaned.
+        """
+
+        if isinstance(input, dict):
+            return self.predict_callable(**input, **(params or {}))
+        else:
```
Comment (B-Step62) [inline mlflow/llama_index/pyfunc_wrapper.py:41]: Can we add explicit `elif` for the other supported type, and raise for unexpected inputs? (or is it complex - like depending on engine type?)

File: mlflow/llama_index/pyfunc_wrapper.py [hunk 34]
```diff
@@ -0,0 +1,159 @@
...
+        super().__init__(*args, **kwargs)
+        self.engine_type = CHAT_ENGINE_NAME
+
+    def _build_engine_method(self) -> Callable:
+        return self.index.as_chat_engine(**self.model_config).chat
+
+    @staticmethod
+    def _parse_dict_as_chat_message_objects(data: dict) -> dict:
+        message_matches = (name for name in _CHAT_MESSAGE_PARAMETER_NAMES if name in data)
+        if (key := next(message_matches, None)) and data.get(key):
```
Comment (B-Step62) [inline mlflow/llama_index/pyfunc_wrapper.py:64]: This part is a bit hard to follow, can we simply construct a for loop?
```
for key in CHAT_MESSAGE_PARAMETER_NAMES:
    if v := data.get(key) and not isinstance(v, str):
```

File: mlflow/llama_index/serialize_objects.py [hunk 35]
```diff
@@ -0,0 +1,183 @@
...
+
+def _sanitize_api_key(object_as_dict: Dict[str, str]) -> Dict[str, str]:
+    keys_to_remove = [k for k in object_as_dict.keys() if "api_key" in k.lower()]
+
+    for k in keys_to_remove:
+        if object_as_dict.pop(k, None):
+            _logger.info(
+                "API key removed from object serialization. At inference time,"
+                " the key must be passed as an environment variable or via inference"
+                " parameters."
```
Comment (B-Step62) [inline mlflow/llama_index/serialize_objects.py:60]: nit: I'd remove `or via inference parameters`.... it will highly likely create a security hole e.g. put API key in the code😅

File: mlflow/llama_index/serialize_objects.py [hunk 36]
```diff
@@ -0,0 +1,183 @@
+import importlib
```
Comment (B-Step62) [inline mlflow/llama_index/serialize_objects.py:1]: Shall we split this out to a separate PR?

File: mlflow/llama_index/pyfunc_wrapper.py (hunk 33 above)
Comment (michael-berk) [reply mlflow/llama_index/pyfunc_wrapper.py:41]: Unfortunately the signature is not consistent, even within a single engine type. I don't see a scalable way to handle unexpected inputs (without essentially recreating python base errors).

File: mlflow/llama_index/pyfunc_wrapper.py (hunk 34 above)
Comment (michael-berk) [reply mlflow/llama_index/pyfunc_wrapper.py:64]: No idea why this is written like this 🙃 

Thanks
